UART_driver/Core/Src/CircularBuffer.cpp

```diff
@@ -0,0 +1,50 @@
+#include "CircularBuffer.hpp"
+#include <cstring>
+
+CircularBuffer::CircularBuffer(uint8_t* buf, uint16_t size) {
+    this->buf = buf;
+    this->size = size;
+}
+
+uint8_t CircularBuffer::peek(uint8_t& res, uint16_t dist) {
+    if(dist + readPtr >= writePtr){
+        return 0;
+    }
+    res = buf[(readPtr + dist) % size];
+    return 1;
+}
+
+uint8_t CircularBuffer::read(uint8_t* res, uint16_t dist) {
+    if(dist + readPtr > writePtr) return 0;
+
+    if( (readPtr % size) + dist >= size ) {
+        // two memcpys needed
+        int dist_to_end = size - (readPtr % size);
+        std::memcpy(res, &buf[readPtr % size], dist_to_end);
+        readPtr += dist_to_end;
+        dist -= dist_to_end;
+        std::memcpy(&res[dist_to_end], &buf[readPtr % size], dist);
+        readPtr += dist;
+    } else {
+        // one memcpy needed
+        std::memcpy(res, &buf[readPtr % size], dist);
+        readPtr += dist;
+    }
+
+    return 1;
+
+}
+
+bool CircularBuffer::hasSpace() {
+    return !((writePtr % size == readPtr % size) && writePtr != readPtr);
+}
+
+uint16_t CircularBuffer::getDataSize() {
+    return writePtr - readPtr;
```

Comment:
assuming this function returns the number of elements currently in the circular buffer? Wouldn't this return bad values sometimes? In the case where the write ptr gets behind the read pointers?

If i mis interpreted what the function does, please disregard.

Reply:
@Chrisc110 We never decrease the read/write pointers, so this works because the write pointer is always >= the read pointer. To index the buffer we just mod the pointers by the buffer size.

The only case this fails is if the pointers exceed the max value of `uint16_t`. Maybe we should decrease the pointers if they're near that limit?

Reply:
Would be a design decision, up to you. What ever you think is reasonable but I don’t foresee a circular buffer needing more than 2^16 elements.

Reply:
I'll leave it as is then :)